### ISSUE TITLE
Flat pattern matching and expressions on indices

### ISSUE THREAD

**Issue by arthuraa** [agda/agda#5501]
Consider the following snippet:

```agda
module Test where

open import Data.Nat

data Bar : ℕ → Set where
  inj₁ : ∀ n → Bar (n + n)
  inj₂ : ∀ n → Bar n

test-bar : (@♭ n : ℕ) → @♭ Bar n → ℕ
test-bar _ (inj₁ n) = n
test-bar _ (inj₂ n) = n
```

If I try to load this in Agda, I get an error:

```
I'm not sure if there should be a case for the constructor inj₁,
because I get stuck when trying to solve the following unification
problems (inferred index ≟ expected index):
  n + n ≟ n₁
Possible reason why unification failed:
  Cannot solve variable n of type ℕ with solution n₁ + n₁ because the
  solution cannot be used at relevant, unrestricted modality
when checking that the pattern inj₁ n has type Bar n
```

Is this a spurious error, or is there a reason behind it? If I replace the expression `n + n` with `n`, the error goes away. On the other hand, if I add other flat declarations in the definition of `Bar` (e.g., when declaring `n`), the error does not go away.

**Comment by andreasabel** [comment]
The reason is the one stated:  Agda cannot find an `n` such that `n + n` is `n_1`.  (And I couldn't either...)

The advice is confine yourself to _patterns_ in indices, and `n + n` is not a pattern because `+` is not a constructor but a defined function.  The standard workaround is to use equality (writing `==` for the triple-dash):
```agda
data Bar : ℕ → Set where
  inj₁ : ∀ n m -> n + n == m → Bar m
...
test-bar _ (inj₁ n m eq) = m
```

**Comment by arthuraa** [comment]
Thanks for the explanation! I think that using an explicit equality might work for me.

However, I still don't see what the issue is. I understand that Agda can run into problems when arbitrary expressions occur as indices of a data type. But here there also seems to be a strange interaction between the indices and the flat modality. E.g., the following modified snippet works even though the constructor arguments are marked as flat:

```agda
module Test where

open import Data.Nat

data Bar : ℕ → Set where
  inj₁ : ∀ (@♭ n) → Bar (n + n)
  inj₂ : ∀ (@♭ n) → Bar n

-- The function arguments are not marked as flat anymore.
test-bar : (n : ℕ) → Bar n → ℕ
test-bar _ (inj₁ n) = n
test-bar _ (inj₂ n) = n
```

Why is the match OK without flat, but yields an error when the arguments are flat?

**Comment by andreasabel** [comment]
A sorry, my explanation was off. 
The error seems to be triggered by a modality mismatch in the unifier, but I do not know the flat modality off the top of my head, so maybe someone else can explain this better, e.g. @Saizan.  In any case the error message could be better.

**Comment by Saizan** [comment]
The main reason we have restrictions on when `@flat` variables can be refined by pattern matching is that local equalities do not generally imply global equalities.

I.e. there are models where `(P : @flat A -> Set) (@flat x  y : A) -> x == y -> P x -> P y` would be unsound, as `x == y` is only talking about x and y being equal as element of A in the default modality, i.e. locally, but not as elements of A at modality flat.

Now though your first example is actually closer to `(P : @flat A -> Set) (@flat x  y : A) -> @flat x == y -> P x -> P y` which would seem sound to me. There might actually be a bug here about propagating the modality of the argument we are matching on.

A possible extension would also be special casing types like `Nat` since most models will have that there's no difference between `(x : Nat)` and `(@flat x : Nat)`.

@arthuraa what is your use case for the flat modality, if you don't mind saying?

**Comment by arthuraa** [comment]
Yes, I am aware that rewriting without the modality would be unsound in some cases (including my own application!). This is why I tried eliminating `Bar` with the modality.

(BTW, is `@0` just an alternative syntax for the flat modality?)

I don't know if this is quite what you meant by your comment, but having special support for types satisfying `@0 T = T` would be nice.

My use case is to model the global sections modality in a presheaf topos to express the elimination principles of higher-order abstract syntax, along the lines sketched by [Hofmann](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.122.6636&rep=rep1&type=pdf). (This might interest @andreasabel.) So far, I've been able to prove the Church-Rosser theorem for the untyped lambda calculus by postulating the type of lambda terms, along with its elimination principle and computation rules. I am trying to understand how much can be done in vanilla Agda without extending the language.

**Comment by Saizan** [comment]
Ah, I have also seen https://github.com/uelis/contextual in that direction, in case you haven't.

Spatial type theory also seems to validate the kind of transports we are talking about, so I'll promote(?) this to a bug.

**Comment by Saizan** [comment]
Btw, no, `@0` is notation for the erasure modality instead, it was a typo in my previous message, I edited it now to `@flat`.

**Comment by arthuraa** [comment]
@Saizan No, I was not aware of this work. Thanks for pointing it out!
